### ISSUE TITLE
duperemove submissions ignored by kernel except changes modify and change times on files submitted for dedup

### ISSUE THREAD

**Issue by dominicraf** [markfasheh/duperemove#51]
using duperemove v0.09.1 with kernel 3.19.3 and options -rdhv I find that after a very long run with many many matched extents found, almost no space is released, presumably because the kernel refuses to honour the requests from duperemove - why? Volume is mounted with options `defaults,compress=lzo,autodefrag` - so compression is used but I believe many of the files refused for deduplication were not in fact compressed.

Even more seriously, all files identified by `duperemove` as having matched extents (except for one file in each case - the first listed by `duperemove -v` I think) have their modify and change times updated. Is this by design? AFAIK the files have not changed in any way, and they have not even been deduped, but the whole system is now wrecked for rsync'ing to non-btrfs fs.

Example extract output from duperemove:

`/home/root-samba/archives/Litos_film.mov       start block: 0 (0)`
`/home/root-samba/archives/rdiff-backup-data/increments/Litos_film.mov.2009-06-03T22:15:03+01:00.snapshot  start block: 0 (0)`
`Dedupe 1 extents with target: (0.0, 225.4M), "/home/root-samba/archives/Litos_film.mov"`
`[...]`
`Kernel processed data (excludes target files): 43.0G`
`Comparison of extent info shows a net change in shared extents of: 128.0K`

Checking afterwards:

`$ sudo stat "/home/root-samba/archives/Litos_film.mov" "/home/root-samba/archives/rdiff-backup-data/increments/Litos_film.mov.2009-06-03T22:15:03+01:00.snapshot"`
`  File: ‘/home/root-samba/archives/Litos_film.mov’`
`  Size: 236343878       Blocks: 461616     IO Block: 4096   regular file`
`Device: 1ah/26d Inode: 14033281    Links: 1`
`Access: (0770/-rwxrwx---)  Uid: ( 1003/dominic-pcchips2)   Gid: (  100/   users)`
`Access: 2015-03-29 14:42:28.025850060 +0100`
`Modify: 2006-12-05 12:31:00.000000000 +0000`
`Change: 2015-02-26 18:27:13.732004959 +0000`
` Birth: -`
`  File: ‘/home/root-samba/archives/rdiff-backup-data/increments/Litos_film.mov.2009-06-03T22:15:03+01:00.snapshot’`
`  Size: 236343878       Blocks: 461616     IO Block: 4096   regular file`
`Device: 1ah/26d Inode: 15861510    Links: 1`
`Access: (0770/-rwxrwx---)  Uid: ( 1008/root-samba)   Gid: ( 1008/root-samba)`
`Access: 2015-03-30 06:49:23.265361050 +0100`
`Modify: 2015-03-30 12:57:55.745174664 +0100`
`Change: 2015-03-30 12:57:55.745174664 +0100`
` Birth: -`
The 2nd file now has modify/change date of 2015-03-30 but in fact has not been changed - except by duperemove - for more than five years.

**Comment by markfasheh** [comment]
Hi, so let me split this into two parts.

1) ctime/mtime

I need to double check in the kernel but the clone operation (which is what happens underneath the dedupe ioctl) is setting mtime/ctime on the files being deduped. Basically it's rewriting file extent pointers and that counts as a change in metadata so you get a ctime update. The mtime update is because clone is doing the work and doesn't know that the data (to the user) hasn't changed. There might be a good case for not doing the mtime update in case of dedupe but I believe we'd always have to do the ctime update because an inode attribute (extent pointer) is changing.


2) You might not be seeing dedupe for a couple reasons - if you see the last two items in the FAQ:

https://github.com/markfasheh/duperemove/blob/master/FAQ.md

either one can lead to this situation but my guess is that you're hitting the second one, a limitation on how btrfs extents. The jist of it is that btrfs won't split underlying extents, even if there's only one reference to it and that reference is to a portion of the extent. Fixing it is something that has to be done but is not easy. I can get into more detail if you like.

In the meantime, would you mind running the 'find-shared-extents' on a file which was selected for dedupe and paste the output to me here? I'd like to see if it has extents marked as shared.

Thanks, and sorry that your rsync got wrecked :( Btw, it's not ideal but you might be able to regain some of that performance using 'rsync -c'

**Comment by dominicraf** [comment]
Thanks Mark for your reply

On 01/04/2015 00:31, markfasheh wrote:
>
> Hi, so let me split this into two parts.
>
> 1) ctime/mtime
>
> I need to double check in the kernel but the clone operation (which is 
> what happens underneath the dedupe ioctl) is setting mtime/ctime on 
> the files being deduped. Basically it's rewriting file extent pointers 
> and that counts as a change in metadata so you get a ctime update. The 
> mtime update is because clone is doing the work and doesn't know that 
> the data (to the user) hasn't changed. There might be a good case for 
> not doing the mtime update in case of dedupe but I believe we'd always 
> have to do the ctime update because an inode attribute (extent 
> pointer) is changing.
>

I think the mtime is more important than the ctime, and it is what rsync 
looks at. If you can stop the mtime update, at least optionally, it 
would be good I think.

> 2) You might not be seeing dedupe for a couple reasons - if you see 
> the last two items in the FAQ:
>
> https://github.com/markfasheh/duperemove/blob/master/FAQ.md
>
> either one can lead to this situation but my guess is that you're 
> hitting the second one, a limitation on how btrfs extents. The jist of 
> it is that btrfs won't split underlying extents, even if there's only 
> one reference to it and that reference is to a portion of the extent. 
> Fixing it is something that has to be done but is not easy. I can get 
> into more detail if you like.
>
> In the meantime, would you mind running the 'find-shared-extents' on a 
> file which was selected for dedupe and paste the output to me here? 
> I'd like to see if it has extents marked as shared.
>
> Thanks, and sorry that your rsync got wrecked :( Btw, it's not ideal 
> but you might be able to regain some of that performance using 'rsync -c'
>

Did you mean 'show-shared-extents'? I tried:
$ sudo show-shared-extents /home
fiemap error 95: Operation not supported

I'm not sure what a fiemap is. Typo maybe?

Regarding rsync, I agree that -c would solve it but I'm also concerned 
that changes to mtime might have other adverse consequences (esp for 
rdiff-backup which I am using). I have managed to revert most of the 
mtimes by picking up old mtimes from a backup and using touch -md on the 
affected files, and I am hoping this will be sufficient.

**Comment by markfasheh** [comment]
Yeah sorry I meant 'show-shared-extents'.

You ran it against a directory which is why it errored. Could you run it against one (or a couple) of the files that duperemove selected for dedupe? In theory that would be any of the ones which gave you the time change.

**Comment by dominicraf** [comment]
Hi Mark
    
    So here are the results for two files (out of many!) which were
    submitted by duperemove as having matching extents (in fact the
    files are duplicates). Hope it helps.
    
    Just to say I think duperemove is amazingly cool, thanks for your
    work on it. If only I could get it behaving the way I need...
    
    Dominic
    
    The duperemove report (extract):
    
    /home/root-samba/archives/goe/Sofia-Jan07/Treality/Litos_film.mov      
    start block: 0 (0)
    /home/root-samba/archives/rdiff-backup-data/increments/goe/Sofia-Jan07/Treality/Litos_film.mov.2009-06-03T22:15:03+01:00.snapshot 
    start block: 0 (0)
    Dedupe 1 extents with target: (0.0, 225.4M),
    "/home/root-samba/archives/goe/Sofia-Jan07/Treality/Litos_film.mov"
    
    
    $ sudo show-shared-extents
    "/home/root-samba/archives/goe/Sofia-Jan07/Treality/Litos_film.mov"
    (fiemap) [0] fe_logical: 0, fe_length: 33554432, fe_physical:
    1734779793408, fe_flags: 0x2000 (shared )
    (fiemap) [1] fe_logical: 33554432, fe_length: 202792960,
    fe_physical: 1689605656576, fe_flags: 0x2001 (last shared )
    /home/root-samba/archives/goe/Sofia-Jan07/Treality/Litos_film.mov:
    236347392 shared bytes
    
    $ sudo show-shared-extents
"/home/root-samba/archives/rdiff-backup-data/increments/goe/Sofia-Jan07/Treality/Litos_film.mov.2009-06-03T22:15:03+01:00.snapshot"
    (fiemap) [0] fe_logical: 0, fe_length: 16777216, fe_physical:
    1734779793408, fe_flags: 0x2000 (shared )
    (fiemap) [1] fe_logical: 16777216, fe_length: 16777216, fe_physical:
    1734796570624, fe_flags: 0x2000 (shared )
    (fiemap) [2] fe_logical: 33554432, fe_length: 16777216, fe_physical:
    1689605656576, fe_flags: 0x2000 (shared )
    (fiemap) [3] fe_logical: 50331648, fe_length: 16777216, fe_physical:
    1689622433792, fe_flags: 0x2000 (shared )
    (fiemap) [4] fe_logical: 67108864, fe_length: 16777216, fe_physical:
    1689639211008, fe_flags: 0x2000 (shared )
    (fiemap) [5] fe_logical: 83886080, fe_length: 16777216, fe_physical:
    1689655988224, fe_flags: 0x2000 (shared )
    (fiemap) [6] fe_logical: 100663296, fe_length: 16777216,
    fe_physical: 1689672765440, fe_flags: 0x2000 (shared )
    (fiemap) [7] fe_logical: 117440512, fe_length: 16777216,
    fe_physical: 1689689542656, fe_flags: 0x2000 (shared )
    (fiemap) [8] fe_logical: 134217728, fe_length: 16777216,
    fe_physical: 1689706319872, fe_flags: 0x2000 (shared )
    (fiemap) [9] fe_logical: 150994944, fe_length: 16777216,
    fe_physical: 1689723097088, fe_flags: 0x2000 (shared )
    (fiemap) [10] fe_logical: 167772160, fe_length: 16777216,
    fe_physical: 1689739874304, fe_flags: 0x2000 (shared )
    (fiemap) [11] fe_logical: 184549376, fe_length: 16777216,
    fe_physical: 1689756651520, fe_flags: 0x2000 (shared )
    (fiemap) [12] fe_logical: 201326592, fe_length: 16777216,
    fe_physical: 1689773428736, fe_flags: 0x2000 (shared )
    (fiemap) [13] fe_logical: 218103808, fe_length: 16777216,
    fe_physical: 1689790205952, fe_flags: 0x2000 (shared )
    (fiemap) [14] fe_logical: 234881024, fe_length: 1441792,
    fe_physical: 1689806983168, fe_flags: 0x2000 (shared )
    (fiemap) [15] fe_logical: 236322816, fe_length: 24576, fe_physical:
    1689808424960, fe_flags: 0x2001 (last shared )
    /home/root-samba/archives/rdiff-backup-data/increments/goe/Sofia-Jan07/Treality/Litos_film.mov.2009-06-03T22:15:03+01:00.snapshot:
    236347392 shared bytes
    
    
    
    On 02/04/2015 00:35, markfasheh wrote:
    
    
      Yeah sorry I meant 'show-shared-extents'.
      You ran it against a directory which is why it errored. Could
        you run it against one (or a couple) of the files that
        duperemove selected for dedupe? In theory that would be any of
        the ones which gave you the time change.
      —
        Reply to this email directly or view
          it on GitHub.

**Comment by markfasheh** [comment]
Thanks for the kind words, they are appreciated! Even if duperemove doesn't work out for you right now this bug will help me understand why so we can catch your use case later.

So from what you're showing me, those files all have their extents marked as shared. Do you have snapshots on that device? If so that will explain the very low estimate duperemove gave you:

"Comparison of extent info shows a net change in shared extents of: 128.0K"

Basically, a snapshot will cause all extents contained in it to be marked as shared. This makes it difficult for duperemove to know how much space was actually saved as there won't be a difference in the flags for the extent after the dedupe.

But otherwise they've been shared with other inodes. The question of how much actual space was saved is better answered with df at this point - do you have an idea at this point as to whether the fs in question shrunk and if so how much?

**Comment by dominicraf** [comment]
Hi Mark
    
    Thanks for the kind words, they are
      appreciated! Even if duperemove doesn't work out for you right now
      this bug will help me understand why so we can catch your use case
      later.
    
    Pleased to be of service!
    
    Re the 'mtime' issue, from what I've now read around (e.g. here
    and here)
    I feel that duperemove should not touch mtime on the files it is
    submitting for deduplication (and nor should btrfs if that is the
    culprit), since mtime changes should reflect changes in the actual
    file not in the metadata. (By contrast ctime should change if
    deduplication takes place since metadata changes [i.e. extents]).
    
    Re the 'no-space-gained' issue, I have looked with `df` and with
    `btrfs fi df` and both return the same info as the (new) `btrfs fi
    usage`, suggesting that deduplication didn't yield any space saving.
    Before running duperemove I had about 100G free. Now I see:
    
    $ sudo df -h /home
    Filesystem                  Size  Used Avail Use% Mounted on
    /dev/mapper/timedicer-home  420G  321G   99G  77% /home
    $ sudo btrfs fi df /home
    Data, single: total=320.00GiB, used=316.23GiB
    System, single: total=32.00MiB, used=44.00KiB
    Metadata, single: total=5.00GiB, used=3.52GiB
    GlobalReserve, single: total=512.00MiB, used=0.00B
    $ sudo btrfs fi usage /home
    Overall:
        Device size:                 420.00GiB
        Device allocated:            325.03GiB
        Device unallocated:           94.97GiB
        Used:                        319.75GiB
        Free (estimated):             98.74GiB      (min: 98.74GiB)
        Data ratio:                       1.00
        Metadata ratio:                   1.00
        Global reserve:              512.00MiB      (used: 0.00B)
    
Data,single: Size:320.00GiB, Used:316.23GiB
       /dev/dm-0     320.00GiB
    
Metadata,single: Size:5.00GiB, Used:3.52GiB
       /dev/dm-0       5.00GiB
    
System,single: Size:32.00MiB, Used:44.00KiB
       /dev/dm-0      32.00MiB
    
Unallocated:
       /dev/dm-0      94.97GiB
    

    Although I don't have the same output from before I ran duperemove,
    I do have a record of space used on the drive shortly beforehand
    which was 319G - and 318G shortly after completing duperemove. This
    all confirms that running duperemove did not save significant space
    on the disk, though I don't understand how this can be - given that
    there is clear evidence (from show-shared-extents) that
    deduplication has occurred and that files are now sharing extents.
    
    Some more information about my system, in case any of it is
    relevant:
    
      No raid
      
      My btrfs volume is mounted over lvm (i.e. inside an lvm
        volume).
      I have no btrfs snapshots (nor any lvm snapshots). 
      
      This is a Vmware virtual machine (i.e. the whole filesystem is
        inside a vmdk file). (I am still using ext4 for the separate
        root filesystem, this is on a separate virtual drive.)
      I have problems balancing this volume, it seems to fail even
        though I am using btrfs-progs v3.19-rc2 with kernel 3.19.3. The
        last time I tried (after using duperemove) it failed.
      
      I am using compression - (mount options:
        defaults,compress=lzo,noatime,autodefrag,subvolid=0), and (just)
        before running duperemove I had 'manually' compressed a lot of
        existing files with btrfs fi defragment -c. This saved about 10G
        space (used space went from 329G to 319G). However at least some
        (and probably most) of the files that were identified by
        duperemove as having shared extents were not defragmented or
        compressed (because duplicate files in my system tend to be
        image files - which I deliberately did not submit for
        defrag/compress).
      
    
    My idea is that btrfs has not released the extents that were freed
    up after deduplication? Is this possible?
    
    I haven't tried `btrfs scrub /home`, maybe I will try that next...
    
    Regards, Dominic
    
    On 03/04/2015 21:28, markfasheh wrote:
    
    
      Thanks for the kind words, they are appreciated! Even if
        duperemove doesn't work out for you right now this bug will help
        me understand why so we can catch your use case later.
      So from what you're showing me, those files all have their
        extents marked as shared. Do you have snapshots on that device?
        If so that will explain the very low estimate duperemove gave
        you:
      "Comparison of extent info shows a net change in shared extents
        of: 128.0K"
      Basically, a snapshot will cause all extents contained in it to
        be marked as shared. This makes it difficult for duperemove to
        know how much space was actually saved as there won't be a
        difference in the flags for the extent after the dedupe.
      But otherwise they've been shared with other inodes. The
        question of how much actual space was saved is better answered
        with df at this point - do you have an idea at this point as to
        whether the fs in question shrunk and if so how much?
      —
        Reply to this email directly or view
          it on GitHub.

**Comment by derobert** [comment]
Also seeing no space freed (according to btrfs usage) after a dedup which should have freed some. Linux 4.0.1 + the revert to fix -m/dconvert.

    Kernel processed data (excludes target files): 41.1G
    Comparison of extent info shows a net change in shared extents of: 0.0

Before:

    Watt:/home/anthony/src/btrfs-progs# ./btrfs fi usage /media/anthony/BigBackup/
    Overall:
        Device size:                   7.28TiB
        Device allocated:              3.35TiB
        Device unallocated:            3.93TiB
        Device missing:                  0.00B
        Used:                          3.34TiB
        Free (estimated):              1.97TiB      (min: 1.97TiB)
        Data ratio:                       2.00
        Metadata ratio:                   2.00
        Global reserve:              512.00MiB      (used: 0.00B)
    
    Data,RAID1: Size:1.67TiB, Used:1.67TiB
       /dev/mapper/luks-562e4e2f-2894-415a-aaf1-7c94a11c33b9           1.67TiB
       /dev/mapper/luks-ec97c1ad-21d8-41bb-9072-e5a74f68e416           1.67TiB
    
    Metadata,RAID1: Size:3.00GiB, Used:1.99GiB
       /dev/mapper/luks-562e4e2f-2894-415a-aaf1-7c94a11c33b9           3.00GiB
       /dev/mapper/luks-ec97c1ad-21d8-41bb-9072-e5a74f68e416           3.00GiB
    
    System,RAID1: Size:32.00MiB, Used:272.00KiB
       /dev/mapper/luks-562e4e2f-2894-415a-aaf1-7c94a11c33b9          32.00MiB
       /dev/mapper/luks-ec97c1ad-21d8-41bb-9072-e5a74f68e416          32.00MiB
    
    Unallocated:
       /dev/mapper/luks-562e4e2f-2894-415a-aaf1-7c94a11c33b9           2.88TiB
       /dev/mapper/luks-ec97c1ad-21d8-41bb-9072-e5a74f68e416           1.06TiB

After

    Watt:/home/anthony/src/btrfs-progs# ./btrfs fi usage /media/anthony/BigBackup/
    Overall:
        Device size:                   7.28TiB
        Device allocated:              3.35TiB
        Device unallocated:            3.93TiB
        Device missing:                  0.00B
        Used:                          3.34TiB
        Free (estimated):              1.97TiB      (min: 1.97TiB)
        Data ratio:                       2.00
        Metadata ratio:                   2.00
        Global reserve:              512.00MiB      (used: 0.00B)
    
    Data,RAID1: Size:1.67TiB, Used:1.67TiB
       /dev/mapper/luks-562e4e2f-2894-415a-aaf1-7c94a11c33b9           1.67TiB
       /dev/mapper/luks-ec97c1ad-21d8-41bb-9072-e5a74f68e416           1.67TiB
    
    Metadata,RAID1: Size:3.00GiB, Used:1.99GiB
       /dev/mapper/luks-562e4e2f-2894-415a-aaf1-7c94a11c33b9           3.00GiB
       /dev/mapper/luks-ec97c1ad-21d8-41bb-9072-e5a74f68e416           3.00GiB
    
    System,RAID1: Size:32.00MiB, Used:272.00KiB
       /dev/mapper/luks-562e4e2f-2894-415a-aaf1-7c94a11c33b9          32.00MiB
       /dev/mapper/luks-ec97c1ad-21d8-41bb-9072-e5a74f68e416          32.00MiB
    
    Unallocated:
       /dev/mapper/luks-562e4e2f-2894-415a-aaf1-7c94a11c33b9           2.88TiB
       /dev/mapper/luks-ec97c1ad-21d8-41bb-9072-e5a74f68e416           1.06TiB

I don't (yet) have the http://permalink.gmane.org/gmane.comp.file-systems.btrfs/44579 patch applied yet, but (if I'm reading the other issue right) that just causes duperemove to report 0 bytes freed.

Running bedup now which, watching the btrfs usage numbers as it runs, is actually freeing space.

**Comment by markfasheh** [comment]
Can you run show-shared-extents in the same way I asked https://github.com/dominicraf to do it? Basically, take a file which you know was 'deduped' and run against it.

If the extents are being marked as shared, most likely, we're hitting the limitation on btrfs extents due to 'bookending': http://en.wikipedia.org/wiki/Btrfs#Extents It's something that I am looking into fixing on btrfs but any patches are a long way off at the moment.


I can't readily explain the difference in dedupe you see from bedupe but it certainly seems worth looking into. If I was to guess with no data, it might be that bedupe in your case found an extent that duperemove threw out (maybe due to extent flags?). Can you tell me perhaps which files bedupe deduped for you? I'd be interested to know if they were compressed, and what show-shared-extents gives us for them.

Thanks,

**Comment by derobert** [comment]
Sorry for the late response.

I made several test files, filled with random data: 12345678 bytes from 
/dev/urandom. I made two of them, file1 and file2. I then copied (with 
plain cp) file1 → file1b and file2 → file2b. Finally, I did `cat file1 
file2 > file12`.

# date; ./duperemove -drh /media/anthony/BigBackup/backup/foo/; date
Wed May 13 02:08:07 EDT 2015
Using 128K blocks
Using hash: SHA256
Using 8 threads for file hashing phase
csum: /media/anthony/BigBackup/backup/foo//file1        [1/5]
csum: /media/anthony/BigBackup/backup/foo//file2        [2/5]
csum: /media/anthony/BigBackup/backup/foo//file1b       [3/5]
csum: /media/anthony/BigBackup/backup/foo//file2b       [4/5]
csum: /media/anthony/BigBackup/backup/foo//file12       [5/5]
Hashed 564 blocks, resulting in 282 unique hashes. Calculating duplicate extents - this may take some time.
[########################################]
Search completed with no errors.
Simple read and compare of file data found 2 instances of extents that might benefit from deduplication.
Start           Length          Filename (3 extents)
0.0     11.8M   "/media/anthony/BigBackup/backup/foo//file1b"
0.0     11.8M   "/media/anthony/BigBackup/backup/foo//file12"
0.0     11.8M   "/media/anthony/BigBackup/backup/foo//file1"
Start           Length          Filename (2 extents)
0.0     11.8M   "/media/anthony/BigBackup/backup/foo//file2"
0.0     11.8M   "/media/anthony/BigBackup/backup/foo//file2b"
Dedupe 2 extents with target: (0.0, 11.8M), "/media/anthony/BigBackup/backup/foo//file1b"
Dedupe 1 extents with target: (0.0, 11.8M), "/media/anthony/BigBackup/backup/foo//file2"
Kernel processed data (excludes target files): 35.2M
Comparison of extent info shows a net change in shared extents of: 35.2M
Wed May 13 02:08:07 EDT 2015

So, first, it missed half of file12—the half that is a duplicate of 
file2. But I guess the block didn't line up. Before duperemove, btrfs fi 
df showed "Data, RAID1: total=1744896.00MiB, used=1683683.00MiB"; 
after... the exact same thing. I waited for over five minutes, there was 
no change (seems btrfs caches free space for a bit).

# ./show-shared-extents /media/anthony/BigBackup/backup/foo/file*
(fiemap) [0] fe_logical: 0, fe_length: 12320768, fe_physical: 6092351303680, fe_flags: 0x2000 (shared )
(fiemap) [1] fe_logical: 12320768, fe_length: 28672, fe_physical: 6090312306688, fe_flags: 0x1 (last )
/media/anthony/BigBackup/backup/foo/file1: 12320768 shared bytes
(fiemap) [0] fe_logical: 0, fe_length: 12320768, fe_physical: 6092351303680, fe_flags: 0x2000 (shared )
(fiemap) [1] fe_logical: 12320768, fe_length: 12374016, fe_physical: 6092388323328, fe_flags: 0x1 (last )
/media/anthony/BigBackup/backup/foo/file12: 24641536 shared bytes
(fiemap) [0] fe_logical: 0, fe_length: 12349440, fe_physical: 6092351303680, fe_flags: 0x2001 (last shared )
/media/anthony/BigBackup/backup/foo/file1b: 36990976 shared bytes
(fiemap) [0] fe_logical: 0, fe_length: 12349440, fe_physical: 6092338954240, fe_flags: 0x2001 (last shared )
/media/anthony/BigBackup/backup/foo/file2: 49340416 shared bytes
(fiemap) [0] fe_logical: 0, fe_length: 12320768, fe_physical: 6092338954240, fe_flags: 0x2000 (shared )
(fiemap) [1] fe_logical: 12320768, fe_length: 28672, fe_physical: 6092375973888, fe_flags: 0x1 (last )
/media/anthony/BigBackup/backup/foo/file2b: 61661184 shared bytes


Next, I ran bedup. It finished at 02:16:06, saying it deduplicated 
file1/1b and file2/2b. Until 02:16:06, the btrfs fi df output stayed the 
same; but at 02:16:39 it changed: Data, RAID1: total=1744896.00MiB, 
used=1683671.25MiB. So bedup managed to free some space (oddly only 12M; 
I'd expect 24M), at least after 30s of waiting or so.

Here is show-shared-extents after bedup:

Watt:/home/anthony/src/github/duperemove# ./show-shared-extents /media/anthony/BigBackup/backup/foo/file*
(fiemap) [0] fe_logical: 0, fe_length: 12320768, fe_physical: 6092351303680, fe_flags: 0x2000 (shared )
(fiemap) [1] fe_logical: 12320768, fe_length: 28672, fe_physical: 6090312306688, fe_flags: 0x2001 (last shared )
/media/anthony/BigBackup/backup/foo/file1: 12349440 shared bytes
(fiemap) [0] fe_logical: 0, fe_length: 12320768, fe_physical: 6092351303680, fe_flags: 0x2000 (shared )
(fiemap) [1] fe_logical: 12320768, fe_length: 12374016, fe_physical: 6092388323328, fe_flags: 0x1 (last )
/media/anthony/BigBackup/backup/foo/file12: 24670208 shared bytes
(fiemap) [0] fe_logical: 0, fe_length: 12320768, fe_physical: 6092351303680, fe_flags: 0x2000 (shared )
(fiemap) [1] fe_logical: 12320768, fe_length: 28672, fe_physical: 6090312306688, fe_flags: 0x2001 (last shared )
/media/anthony/BigBackup/backup/foo/file1b: 37019648 shared bytes
(fiemap) [0] fe_logical: 0, fe_length: 12349440, fe_physical: 6092338954240, fe_flags: 0x2001 (last shared )
/media/anthony/BigBackup/backup/foo/file2: 49369088 shared bytes
(fiemap) [0] fe_logical: 0, fe_length: 12349440, fe_physical: 6092338954240, fe_flags: 0x2001 (last shared )
/media/anthony/BigBackup/backup/foo/file2b: 61718528 shared bytes

There aren't any funny mount options:

$ grep BigBackup /proc/mounts
/dev/mapper/luks-ec97c1ad-21d8-41bb-9072-e5a74f68e416 /media/anthony/BigBackup btrfs rw,nosuid,nodev,relatime,space_cache 0 0

`rm` of all the test files freed another 59MB (again after ≈30s).

**Comment by markfasheh** [comment]
Ok, I want to figure out what version of bedupe you are using - is it using the dedupe syscall or the clone syscall? If it's doing clone then that would point to an issue with the dedupe ioctl (which i can look into).

Easiest would be for you to tell me what your top commit id is. Also, what branch you're on (if you're on dedupe-ioctl then the method is obvious).

**Comment by derobert** [comment]
On Wed, May 20, 2015 at 02:17:27PM -0700, markfasheh wrote:
> Ok, I want to figure out what version of bedupe you are using - is it using the dedupe syscall or the clone syscall? If it's doing clone then that would point to an issue with the dedupe ioctl (which i can look into).

I'm pretty sure my bedup was built from master, most recent commit
5189e166145b8954ac41883f81ef3c3b50dc96ab

Here is the first few lines from "git log --pretty=oneline":

5189e166145b8954ac41883f81ef3c3b50dc96ab Use select_entity_from for SQLAlchemy 0.9+ compatibility
80cb217d4819a03e159e42850a9a3f14e2b278a3 Hints for reporting bugs
244cc49b779a81f2dc1bf1df62cbecc53e39d454 tox: Run tests with fatal warnings
b9b9ddeab426cc35e7cdfcb51b5ee70c6aba7410 Don't use %s on bytes
6fb1c7599af825537765df3ba5942fba18a6427e tox: Revert to released versions
d06f69774c0812f0b4e247232da49fb6462c9ac9 3.10 is out
5fc7e723bf362376778ad7dbdef16576c5370246 Update SA reflection
94fc498fcf1f916eed38f2e4c0cee18a18c98eea size-lookup: handle unavailable filesystems
53dcce88c18fc8b703220358d3151c38a21c1335 Rework pre-migration compatibility

**Comment by dominicraf** [comment]
Like @derobert I also had success with bedup (before I tried duperemove), I did not use the 'experimental' bedup branch so I think the version I had was using the 'old' clone ioctl.

**Comment by markfasheh** [comment]
Ok, this all points to something in the dedupe ioctl, I will investigate and patch it up once it's been figured out.

**Comment by dominicraf** [comment]
The problem seems to be that although btrfs is sharing extents correctly it is not releasing the now-unused duplicate extents. Doing a scrub does not seem to help. Hope you manage to catch the critter! The mtime issue remains a deal-breaker for me too.

**Comment by markfasheh** [comment]
You are correct - if I make two identical files, run show-shared-extents before, then run duperemove to dedupe them the show-shared-extents afterwards will show that the extents are now shared and also will now have the same physical start offset (showing they are the same exact extent). This is on the latest git kernel, which also has my fiemap fix. It's not immediately obvious why the space isn't being freed so it'll take a bit to figure out.

**Comment by markfasheh** [comment]
Ok so this turned out to be an alignment issue, requiring a fix in kernel and a fix in userspace.

The duperemove fix is in master branch, and the kernel fix is queued up for the 4.2 kernel in the btrfs tree:

https://git.kernel.org/cgit/linux/kernel/git/mason/linux-btrfs.git/commit/?h=integration-4.2&id=e1d227a42ea2b4664f94212bd1106b9a3413ffb8


I have a followup patch series in review as well, which fixes more bugs and adds two features, one of which is to skip changing mtime as we discussed previously in this issue.

http://www.spinics.net/lists/linux-btrfs/msg44956.html

**Comment by dominicraf** [comment]
That's great, thanks for your work, including the mtime fix. Can you 
think of a way to recover the lost space on existing btrfs volumes from 
earlier faulty dedupes?

**Comment by markfasheh** [comment]
I'm pretty sure that the updated duperemove would pick up the leftover tail extents and dedupe them. It might require a switch to turn off some logic that tracks already-deduped extents. I'll try to give you a more detailed answer (or put it in the FAQ) in the coming days.

**Comment by 80tf3262p8** [comment]
Thank you, markfasheh. It's especially good to hear the mtime issue will be resolvable. My only confusion is to why bedup reportedly was able to free the unused extents without a kernel update. If a kernel issue is involved, do you think for users who've run bedup, it's only a matter of time until problems appear? If not, do you think it would be feasible to emulate bedup's behavior for systems without the kernel patch, or should we just use bedup until we're able to update?

**Comment by markfasheh** [comment]
Bedup is using the clone ioctl directly. Duperemove uses the dedupe ioctl which uses the clone code as well. The difference is that dedupe was rejecting some unaligned extent lengths whereas clone would be fine with those. So that explains why bedup worked for you. Basically dedupe was too strict with what it would accept, whereas clone is pretty liberal.

The downside is that clone is destructive in that it won't check the targets data. So if you race writing some data to a file that you're cloning to, you could lose that data. The dedupe ioctl checks file data before allowing the extent to be overwritten, therefore you do not have that possibility of losing data.

I don't want to emulate bedups behavior in duperemove because one promise we make is that dedupe won't (to our knowledge) corrupt user data and the dedupe ioctl is the only way to be sure of that.

As far as when you'll get the patches (which I think is the real question here), once they're all in Linus tree (or I'm convinced they'll wind up there) I'd like to make them available, with description in the duperemove tree so it's easy for people who compile their own kernel to get them. I'll also try to get the most important ones backported to the older stable kernel releases.

Lastly if you're running SUSE or openSUSE I can have a more direct role in getting the patches into those kernels so you can expect them their sooner rather than later. That isn't an advertisment or anything btw, I just happen to work there so I can push the patches myself.

**Comment by markfasheh** [comment]
@dominicraf so the (hopefully) final answer to "what do i do if i have these undeduped tails" is to just run duperemove on the files again.

You can run either master, v0.09 or whatever the next v0.10 beta will be.

Master branch will catch this situation for us as of commit f239c925b43127eb34557800b89f38fb38681955 (to be ported to v0.10 branch)

v0.09 doesn't try to figure out if an extent was already deduped, and I have backported the code to submit these to it. So as long as you run v0.09.5 or higher, you should be good as it will always find them and submit the entire extent.

**Comment by dominicraf** [comment]
Thanks Mark. I now have kernel 4.1.1 (and duperemove 0.09.5). Will this work or do I need to wait for kernel 4.2 (and your confirmation that the fix is in there)?

**Comment by markfasheh** [comment]
Yeah you're going to need to either wait for kernel 4.2 or grab the backported fixes from my github:

https://github.com/markfasheh/linux-4.1-dedupe_fixes


I will see about having at least the bug fixes backported to -stable (also others can request this on the btrfs mailing list)

**Comment by olifre** [comment]
Thanks for the good work! I just saw the patches seem to have landed in the fresh 4.2 release. Will current duperemove master already make use of suppressing the mtime-update on clone by default, or is there something special one needs to do?

**Comment by markfasheh** [comment]
Yeah the change is entirely kernel-side so as long as you've updated the kernel you should be seeing the better mtime behavior.

**Comment by lpirl** [comment]
Thanks @markfasheh ! Great work.
